In [1]:
import pandas as pd
import env
import seaborn as sns
import acquire
import prepare
import numpy as np
import matplotlib.pyplot as plt
import scipy
from sklearn.model_selection import train_test_split
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import LabelEncoder, OneHotEncoder, MinMaxScaler
import warnings
warnings.filterwarnings("ignore")

# Exercise 1

Given the following confusion matrix, evaluate (by hand) the model's performance.


|               | actual cat | actual dog |
|:------------  |-----------:|-----------:|
| predicted cat |         34 |          7 |
| predicted dog |         13 |         46 |

- In the contextof this problem, what is a false positive?

- In the context of this problem, what is a false negative?

- TP: predicted cat, actual is cat (34)
- FP: predicted cat, actual is *__not cat__* (7)
- FN: predicted *__not cat__*, actual is cat (13)
- TN: predicted  *__not cat__*, actual is  *__not cat__* (46)

## Metrics 
- __accuracy__: how well the model does overall
> __(TP + TN) / (TP + TN + FP + FN)__
(34 + 46) / (34 + 7 + 13 +46) = 80.0%
- __precision__: % of actually positive cases that were predicted as positive
> __TP / (TP + FP)__
34 / (34 + 7) = 82.9%
- __recall__: % of of positive predictions that are correct
> __TP / (TP + FN)__
34 / (34 + 13) = 72.2%

- How would you describe this model?
> The model does a good job of predicting hits

# Exercise 2
#### You are working as a datascientist working for Codeup Cody Creator (C3 for short), a rubber-duck manufacturing plant.

- Unfortunately, some of the rubber ducks that are produced will have defects. Your team has built several models that try to predict those defects, and the data from their predictions can be found here.

In [2]:
# Read in csv
df = pd.read_csv('c3.csv')

In [3]:
# Peak at the data
df.head()

actual     model1  model2     model3
0  No Defect  No Defect  Defect  No Defect
1  No Defect  No Defect  Defect     Defect
2  No Defect  No Defect  Defect  No Defect
3  No Defect     Defect  Defect     Defect
4  No Defect  No Defect  Defect  No Defect

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 4 columns):
actual    200 non-null object
model1    200 non-null object
model2    200 non-null object
model3    200 non-null object
dtypes: object(4)
memory usage: 6.4+ KB


In [5]:
df.nunique()

actual    2
model1    2
model2    2
model3    2
dtype: int64

- I know that:
> - The data has no null values
> - The data types are all as expected
> - There are 200 observations
> - There are only 2 possible outcomes: __Defect and No defect__
- Awesome! Let's add in a baseline to compare our models to. I'll start by figuring out which observation occurs the most

In [6]:
df.actual.value_counts()

No Defect    184
Defect        16
Name: actual, dtype: int64

In [7]:
# add a baseline predictions columns to the df
df['baseline'] = 'No Defect'
# Verify that all the observations in the new column are 'No Defect'
df.baseline.value_counts()

No Defect    200
Name: baseline, dtype: int64

In [8]:
df.head()

actual     model1  model2     model3   baseline
0  No Defect  No Defect  Defect  No Defect  No Defect
1  No Defect  No Defect  Defect     Defect  No Defect
2  No Defect  No Defect  Defect  No Defect  No Defect
3  No Defect     Defect  Defect     Defect  No Defect
4  No Defect  No Defect  Defect  No Defect  No Defect

In [9]:
# Determine accuracy, precision and recall for model1
positive = 'Defect'

# accuracy -- overall hit rate
model_accuracy = (df.model1 == df.actual).mean()
baseline_accuracy = (df.baseline == df.actual).mean()

# precision -- how good are our positive predictions?
# precision -- model performance | predicted positive
subset = df[df.model1 == positive]
model_precision = (subset.model1 == subset.actual).mean()
subset = df[df.baseline == positive]
baseline_precision = (subset.baseline == subset.actual).mean()

# recall -- how good are we at detecting actual positives?
# recall -- model performance | actual positive
subset = df[df.actual == positive]
model_recall = (subset.model1 == subset.actual).mean()
baseline_recall = (subset.baseline == subset.actual).mean()


print(f'   model accuracy: {model_accuracy:.2%}')
print(f'baseline accuracy: {baseline_accuracy:.2%}')
print()

print(f'model precision: {model_precision:.2%}')
print(f'baseline precision: {baseline_precision:.2%}')
print()

print(f'   model recall: {model_recall:.2%}')
print(f'baseline recall: {baseline_recall:.2%}')


   model accuracy: 95.00%
baseline accuracy: 92.00%

model precision: 80.00%
baseline precision: nan%

   model recall: 50.00%
baseline recall: 0.00%


In [10]:
# Determine accuracy, precision and recall for model2
positive = 'Defect'

# accuracy -- overall hit rate
model_accuracy = (df.model2 == df.actual).mean()
baseline_accuracy = (df.baseline == df.actual).mean()

# precision -- how good are our positive predictions?
# precision -- model performance | predicted positive
subset = df[df.model2 == positive]
model_precision = (subset.model2 == subset.actual).mean()
subset = df[df.baseline == positive]
baseline_precision = (subset.baseline == subset.actual).mean()

# recall -- how good are we at detecting actual positives?
# recall -- model performance | actual positive
subset = df[df.actual == positive]
model_recall = (subset.model2 == subset.actual).mean()
baseline_recall = (subset.baseline == subset.actual).mean()


print(f'   model accuracy: {model_accuracy:.2%}')
print(f'baseline accuracy: {baseline_accuracy:.2%}')
print()

print(f'model precision: {model_precision:.2%}')
print(f'baseline precision: {baseline_precision:.2%}')
print()

print(f'   model recall: {model_recall:.2%}')
print(f'baseline recall: {baseline_recall:.2%}')

   model accuracy: 56.00%
baseline accuracy: 92.00%

model precision: 10.00%
baseline precision: nan%

   model recall: 56.25%
baseline recall: 0.00%


In [11]:
# Determine accuracy, precision and recall for model3
positive = 'Defect'

# accuracy -- overall hit rate
model_accuracy = (df.model3 == df.actual).mean()
baseline_accuracy = (df.baseline == df.actual).mean()

# precision -- how good are our positive predictions?
# precision -- model performance | predicted positive
subset = df[df.model3 == positive]
model_precision = (subset.model3 == subset.actual).mean()
subset = df[df.baseline == positive]
baseline_precision = (subset.baseline == subset.actual).mean()

# recall -- how good are we at detecting actual positives?
# recall -- model performance | actual positive
subset = df[df.actual == positive]
model_recall = (subset.model3 == subset.actual).mean()
baseline_recall = (subset.baseline == subset.actual).mean()


print(f'   model accuracy: {model_accuracy:.2%}')
print(f'baseline accuracy: {baseline_accuracy:.2%}')
print()

print(f'model precision: {model_precision:.2%}')
print(f'baseline precision: {baseline_precision:.2%}')
print()

print(f'   model recall: {model_recall:.2%}')
print(f'baseline recall: {baseline_recall:.2%}')

   model accuracy: 55.50%
baseline accuracy: 92.00%

model precision: 13.13%
baseline precision: nan%

   model recall: 81.25%
baseline recall: 0.00%


- Use the predictions dataset and pandas to help answer the following questions:

> An internal team wants to investigate the cause of the manufacturing defects. They tell you that they want to identify as many of the ducks that have a defect as possible. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?
- positive is __defective__
>> - TP: predicted defective, actual is deffecitve
>> - FP: predicted defective, actual is *__not deffective__*
>> - FN: predicted *__not deffective__*, actual is deffective
>> - TN: predicted  *__not deffectice__*, actual is  *__not deffective__*
- If the goal is to catch as many defective ducks as possible, the cost of a false positive (we predict it's defective but it"s actually not deffective) is not very high.
- The cost of a false negative (we predict it is not deffective when it really is) is high.
- *__Recall__* would be the best metric as it will optimize a reduction in false negatives.
- The best model to use is model 3 as it has the highest recall among all the models including the baseline at: 81.25%

> Recently several stories in the local news have come out highlighting customers who received a rubber duck with a defect, and portraying C3 in a bad light. The PR team has decided to launch a program that gives customers with a defective duck a vacation to Hawaii. They need you to predict which ducks will have defects, but tell you the really don't want to accidentally give out a vacation package when the duck really doesn't have a defect. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?
- If the emphasis is on not giving out expensive vacations, the cost of a false positive (we predict it's deffective but it's actually not deffective) is high.
- *__Precision__* would be the best metric as it will optimize a reduction in false positives.
- The best model to use is model 1 as it has the highest precision among all the models including the baseline at: 80.00%

# Exercise 3

### You are working as a data scientist for Gives You Paws ™, a subscription based service that shows you cute pictures of dogs or cats (or both for an additional fee).

- At Gives You Paws, anyone can upload pictures of their cats or dogs. The photos are then put through a two step process:
> - First an automated algorithm tags pictures as either a cat or a dog (Phase I). 
> - Next, the photos that have been initially identified are put through another round of review, possibly with some human oversight, before being presented to the users (Phase II).

- Several models have already been developed with the data, and you can find their results here.

Given this dataset, use pandas to create a baseline model (i.e. a model that just predicts the most common class) and answer the following questions:

- TP: predicted dog, actual is dog
- FP: predicted dog, actual is *__cat__*
- FN: predicted *__cat__*, actual is dog
- TN: predicted  *__cat__*, actual is  *__cat__*

In [12]:
# Read in csv
df = pd.read_csv('gives_you_paws.csv')

In [13]:
# peak at the data
df.head()

actual model1 model2 model3 model4
0    cat    cat    dog    cat    dog
1    dog    dog    cat    cat    dog
2    dog    cat    cat    cat    dog
3    dog    dog    dog    cat    dog
4    cat    cat    cat    dog    dog

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 5 columns):
actual    5000 non-null object
model1    5000 non-null object
model2    5000 non-null object
model3    5000 non-null object
model4    5000 non-null object
dtypes: object(5)
memory usage: 195.4+ KB


In [15]:
df.nunique()

actual    2
model1    2
model2    2
model3    2
model4    2
dtype: int64

- I know that:
> - The data has no null values
> - The data types are all as expected
> - There are 5000 observations
> - There are only 2 possible outcomes: __cat and dog__
- Awesome! Let's add in a baseline to compare our models to. I'll start by figuring out which observation occurs the most

In [16]:
df.actual.value_counts()

dog    3254
cat    1746
Name: actual, dtype: int64

In [17]:
# add a baseline predictions columns to the df
df['baseline'] = 'dog'
# Verify that all the observations in the new column are 'No Defect'
df.baseline.value_counts()

dog    5000
Name: baseline, dtype: int64

In [18]:
df.head()

actual model1 model2 model3 model4 baseline
0    cat    cat    dog    cat    dog      dog
1    dog    dog    cat    cat    dog      dog
2    dog    cat    cat    cat    dog      dog
3    dog    dog    dog    cat    dog      dog
4    cat    cat    cat    dog    dog      dog

In [19]:
# Determine accuracy, precision and recall for model1
positive = 'dog'

# accuracy -- overall hit rate
model1_accuracy = (df.model1 == df.actual).mean()
baseline_accuracy = (df.baseline == df.actual).mean()

# precision -- how good are our positive predictions?
# precision -- model performance | predicted positive
subset = df[df.model1 == positive]
model1_precision = (subset.model1 == subset.actual).mean()
subset = df[df.baseline == positive]
baseline_precision = (subset.baseline == subset.actual).mean()

# recall -- how good are we at detecting actual positives?
# recall -- model performance | actual positive
subset = df[df.actual == positive]
model1_recall = (subset.model1 == subset.actual).mean()
baseline_recall = (subset.baseline == subset.actual).mean()


print(f'   model accuracy: {model1_accuracy:.2%}')
print(f'baseline accuracy: {baseline_accuracy:.2%}')
print()

print(f'model precision: {model1_precision:.2%}')
print(f'baseline precision: {baseline_precision:.2%}')
print()

print(f'   model recall: {model1_recall:.2%}')
print(f'baseline recall: {baseline_recall:.2%}')

   model accuracy: 80.74%
baseline accuracy: 65.08%

model precision: 89.00%
baseline precision: 65.08%

   model recall: 80.33%
baseline recall: 100.00%


In [20]:
# Determine accuracy, precision and recall for model2
positive = 'dog'

# accuracy -- overall hit rate
model2_accuracy = (df.model2 == df.actual).mean()
baseline_accuracy = (df.baseline == df.actual).mean()

# precision -- how good are our positive predictions?
# precision -- model performance | predicted positive
subset = df[df.model2 == positive]
model2_precision = (subset.model2 == subset.actual).mean()
subset = df[df.baseline == positive]
baseline_precision = (subset.baseline == subset.actual).mean()

# recall -- how good are we at detecting actual positives?
# recall -- model performance | actual positive
subset = df[df.actual == positive]
model2_recall = (subset.model2 == subset.actual).mean()
baseline_recall = (subset.baseline == subset.actual).mean()


print(f'   model accuracy: {model2_accuracy:.2%}')
print(f'baseline accuracy: {baseline_accuracy:.2%}')
print()

print(f'model precision: {model2_precision:.2%}')
print(f'baseline precision: {baseline_precision:.2%}')
print()

print(f'   model recall: {model2_recall:.2%}')
print(f'baseline recall: {baseline_recall:.2%}')

   model accuracy: 63.04%
baseline accuracy: 65.08%

model precision: 89.32%
baseline precision: 65.08%

   model recall: 49.08%
baseline recall: 100.00%


In [21]:
# Determine accuracy, precision and recall for model3
positive = 'dog'

# accuracy -- overall hit rate
model3_accuracy = (df.model3 == df.actual).mean()
baseline_accuracy = (df.baseline == df.actual).mean()

# precision -- how good are our positive predictions?
# precision -- model performance | predicted positive
subset = df[df.model3 == positive]
model3_precision = (subset.model3 == subset.actual).mean()
subset = df[df.baseline == positive]
baseline_precision = (subset.baseline == subset.actual).mean()

# recall -- how good are we at detecting actual positives?
# recall -- model performance | actual positive
subset = df[df.actual == positive]
model3_recall = (subset.model3 == subset.actual).mean()
baseline_recall = (subset.baseline == subset.actual).mean()


print(f'   model accuracy: {model3_accuracy:.2%}')
print(f'baseline accuracy: {baseline_accuracy:.2%}')
print()

print(f'model precision: {model3_precision:.2%}')
print(f'baseline precision: {baseline_precision:.2%}')
print()

print(f'   model recall: {model3_recall:.2%}')
print(f'baseline recall: {baseline_recall:.2%}')

   model accuracy: 50.96%
baseline accuracy: 65.08%

model precision: 65.99%
baseline precision: 65.08%

   model recall: 50.86%
baseline recall: 100.00%


In [22]:
# Determine accuracy, precision and recall for model4
positive = 'dog'

# accuracy -- overall hit rate
model4_accuracy = (df.model4 == df.actual).mean()
baseline_accuracy = (df.baseline == df.actual).mean()

# precision -- how good are our positive predictions?
# precision -- model performance | predicted positive
subset = df[df.model4 == positive]
model4_precision = (subset.model4 == subset.actual).mean()
subset = df[df.baseline == positive]
baseline_precision = (subset.baseline == subset.actual).mean()

# recall -- how good are we at detecting actual positives?
# recall -- model performance | actual positive
subset = df[df.actual == positive]
model4_recall = (subset.model4 == subset.actual).mean()
baseline_recall = (subset.baseline == subset.actual).mean()


print(f'   model accuracy: {model4_accuracy:.2%}')
print(f'baseline accuracy: {baseline_accuracy:.2%}')
print()

print(f'model precision: {model4_precision:.2%}')
print(f'baseline precision: {baseline_precision:.2%}')
print()

print(f'   model recall: {model4_recall:.2%}')
print(f'baseline recall: {baseline_recall:.2%}')

   model accuracy: 74.26%
baseline accuracy: 65.08%

model precision: 73.12%
baseline precision: 65.08%

   model recall: 95.57%
baseline recall: 100.00%


- In terms of accuracy, how do the various models compare to the baseline model? Are any of the models better than the baseline?
> Model 1 has the best accuracy overall when compared to the baseline model
- Suppose you are working on a team that solely deals with dog pictures. Which of these models would you recomend for Phase I? For Phase II?
> - For phase 1 I care more about not catching the dog photo than tagging something as a dog when it's not so I want high *__Recall__*
> - For phase 2 I want to be very sure about positively identifying the dog photos so I want high *__Precision__*
- Suppose you are working on a team that solely deals with cat pictures. Which of these models would you recomend for Phase I? For Phase II?
> - For phase 1 I want to be really sure this is not a dog so I want high *__Precision__* 
> - For phase 2 false positives don't cost as much so I want really high *__Recall__*

# Exercise 4

Follow the links below to read the documentation about each function, then apply those functions to the data from the previous problem.

- [accuracy](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.accuracy_score.html)
- [precision](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.precision_score.html)
- [recall](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.recall_score.html)
- [classification report](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.classification_report.html)